### ISSUE TITLE
Spike: Migrate MessageSubscriptionState to RocksDB

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#1291]
* evaluate how RocksDB iterators work (and sorting columns and column families)
 * find a solution how to store dynamic values into rocksdb, without reallocating every time. Example: layer above rocksDB, which uses an ExpandableBuffer. Layers supports methods for direct buffer so the usage will be simplified for us

**Comment by Zelldon** [comment]
Tested rocks db functionality, to find out how iterators and sorting works.
See the following snipped for detailed information. 
Iterating over key-value pairs returns them in sorted (asc) order.
It is possible to create so called column families and iterate only over them.
Via iterating it is also possible to seek to a specific prefix. If key already exist, then value will be overwritten as expected.

```java

public class StateControllerTest {
  @Rule public TemporaryFolder folder = new TemporaryFolder();

  private StateController stateController;

  @Before
  public void setUp() throws Exception {
    stateController = new StateController();
    stateController.open(folder.newFolder("rocksdb"), false);
  }

  @After
  public void close() {
    stateController.close();
  }

  @Test
  public void shouldIterateAsc() {
    // given
    // when
    stateController.put("foobar".getBytes(), "bar".getBytes());
    stateController.put("abc".getBytes(), "123".getBytes());
    stateController.put("foo".getBytes(), "lulz".getBytes());
    stateController.put("123".getBytes(), "lulz".getBytes());

    // then
    final List<Tuple<byte[], byte[]>> keyValuePairs = new ArrayList<>();
    stateController.foreach(
        (key, value) -> {
          keyValuePairs.add(new Tuple<>(key, value));
        });

    assertThat(keyValuePairs)
        .extracting(t -> t.getLeft())
        .containsExactly("123".getBytes(), "abc".getBytes(), "foo".getBytes(), "foobar".getBytes());
  }

  @Test
  public void shouldIterateOnlyValuesFromColumnFamily() throws Exception {
    // given
    final ColumnFamilyDescriptor descriptor = new ColumnFamilyDescriptor("foo".getBytes());
    final ColumnFamilyHandle columnFamilyHandle =
        stateController.getDb().createColumnFamily(descriptor);

    // when
    stateController.put("abc".getBytes(), "123".getBytes());
    stateController.put("foobar".getBytes(), "bar".getBytes());
    stateController.put("foo".getBytes(), "lulz".getBytes());
    stateController.getDb().put(columnFamilyHandle, "123".getBytes(), "lulz".getBytes());

    // then
    final List<Tuple<byte[], byte[]>> keyValuePairs = new ArrayList<>();
    stateController.foreach(
        columnFamilyHandle,
        (key, value) -> {
          keyValuePairs.add(new Tuple<>(key, value));
        });

    assertThat(keyValuePairs).extracting(t -> t.getLeft()).containsExactly("123".getBytes());
  }

  @Test
  public void shouldOverwrite() throws Exception {
    // given
    // when
    final byte[] keyBytes = "abc".getBytes();
    stateController.put(keyBytes, "123".getBytes());
    stateController.put(keyBytes, "456".getBytes());

    // then
    assertThat(stateController.getDb().get(keyBytes)).containsExactly("456".getBytes());
  }

  @Test
  public void shouldIterateSubset() {
    // given
    stateController.put("abc".getBytes(), "bar".getBytes());
    stateController.put("name13223".getBytes(), "bar".getBytes());
    stateController.put("name123".getBytes(), "bar".getBytes());
    stateController.put("namecorrelation123".getBytes(), "bar".getBytes());
    stateController.put("namecorrelation2323".getBytes(), "bar".getBytes());
    stateController.put("namecorrelation3243".getBytes(), "bar".getBytes());
    stateController.put("namecorrelation14223".getBytes(), "bar".getBytes());

    // when
    final List<Tuple<byte[], byte[]>> keyValuePairs = new ArrayList<>();
    stateController.foreach(
        "namecorrelation".getBytes(),
        (key, value) -> {
          keyValuePairs.add(new Tuple<>(key, value));
        });

    // then
    assertThat(keyValuePairs)
        .extracting(t -> t.getLeft())
        .containsExactly(
            "namecorrelation123".getBytes(),
            "namecorrelation14223".getBytes(),
            "namecorrelation2323".getBytes(),
            "namecorrelation3243".getBytes());
  }
}
```

Created implementation of `MessageStateController`.
Uses column families for different concerns, like time to live or simply existence of message id. 
With this it is possible to iterate over message where deadline is before a given time and check existence for message id. Message is stored with composite key (name and correlation key).
To fix #1205 used reflection to call internal methods.

**Comment by Zelldon** [comment]
In the POC we tried to create a state controller, which should work with the current message stream processors.
 This means it should be possible to store messages, message subscriptions and the storage should work as before. 
 To map the current data store and his use cases to the new state controller, we separated the stored content in
 different column families.

 The different column families are:

  * default: used to store each message (key: name-correlationkey value: message)
  * timeToLive: used to find messages which deadline are due (key consist of time-name-correlationkey)
  * messageId: used to check existence of message with given id (key consist of name-correlationkey-id)
  * subscription: used to store the message subscriptions (key: partitionId-wfInstancekey-actInstanceKey)
  * subSendTime: used to find message subscriptions which are out dated (key: sendTime-partitionId-wfInstancekey-actInstanceKey)

 On adding a message the message will be stored in two or three (if message has an id) different column families.
 The state controller offers a method to find all messages which deadline are due, for this the timeToLive column family is used.
 It is also possible to check the existence of an message, if the message has an specific message id or to simply read the message from
 rocks db via the message name and correlation key.

 On challenge/problem with this approach is, that old store uses a simple list of messages instead of a map. It could happen that the list contains messages with the same name and correlation key.
 If we have now two messages with the same name and correlation key, they will overwrite each other, even if they have 
 different message ids. This could be also problem on removing a message, since we have to clean up all column families. (orphan entries in timeToLive and messageId family)

 On the old implementation messages are removed via the messages key. This means the list was iterated and only the message with the specific key 
 was removed. With the new state controller this would not be longer the case, since we only have one message with the same name and correlation key in the map 
 we simply use this to remove the message. But this could cause some race conditions. It can happen that the DELETE command, which removes the message, comes
 after a second publish of a message with the same name and correlation key. So the message was updated (deadline etc.) but was deleted directly afterwards. To
 avoid that, we could check if the deadline is smaller then the current time.
 
 So there is the open question: Is it a problem, if message with the same name and correlation key overwrite each other? I assume not.

**Comment by saig0** [comment]
> So there is the open question: Is it a problem, if message with the same name and correlation key overwrite each other? I assume not.

Actually, this is a problem :sweat_smile: It should be possible to publish multiple messages with the same name and correlation key. Currently, you loose these messages if the subscription is not opened yet.

**Comment by Zelldon** [comment]
But how? I mean even if another message comes in a message with the same name and key is still there so no lost?

**Comment by saig0** [comment]
Currently, the messages are stored in the `MessageStateController` under the composite key <name, correlation-key>. So a second message with the same name and correlation key overrides the previous one. In the following situation, the first message is lost (i.e. not correlated):

1) publish message 1 (name: ping, correlation-key: pong, payload: foo)
2) publish message 2 (name: ping, correlation-key: pong, payload: bar)
3) workflow instance enteres message catch event  (name: ping, correlation-key: pong)
4) message 2 is correlated (instead of message 1)

When the subscription is opened then the stream processor correlate the first message with the name and correlation key.

**Comment by Zelldon** [comment]
Yes, because name and key was assumed as identifier, so for me this means a message with the same composite key is the same. If this is not the case then we need a composite key of name, correlation key and message event key. But how then the correlation should work? The first one is correlated, which is found with the name and correlation key? This is how it works before. Looks a bit weird for me. I send two messages with the same key and name and the first is correlated to the next subscription which is opened ? 😅

**Comment by Zelldon** [comment]
But if you say this is the way how it should work ok then please create an issue. I will fix it next week.

**Comment by saig0** [comment]
Yes, this was my plan how it should work :smile: 

For me, it's like a queue. If you publish a message then it is inserted into the queue. When the workflow instance opens a subscription then it polls the first message from this queue. If the instance opens a subscription again then it polls the second one.

Does this makes sense for you? We can discuss the concept next week ;)
I'll create an issue and maybe work on it this week.

**Comment by Zelldon** [comment]
Ok with the concept of a queue this makes sense you're right. We should definitely document this properly 😁

**Comment by saig0** [comment]
The state will be fixed by #1405.
